packages/analytics-core/src/plugins/destination.ts

```diff
@@ -297,21 +296,33 @@ export class Destination implements DestinationPlugin {
   }
 
   fulfillRequest(list: Context[], code: number, message: string) {
-    this.saveEvents();
+    this.updateEventStorage(list);
     list.forEach((context) => context.callback(buildResult(context.event, code, message)));
   }
 
   /**
-   * Saves events to storage
    * This is called on
    * 1) new events are added to queue; or
    * 2) response comes back for a request
+   *
+   * Update the event storage based on the queue
    */
-  saveEvents() {
+  updateEventStorage(eventsToRemove?: Context[]) {
     if (!this.config.storageProvider) {
       return;
     }
-    const events = Array.from(this.queue.map((context) => context.event));
-    void this.config.storageProvider.set(this.storageKey, events);
+
+    const eventsToRemoveInsertIdSet = eventsToRemove?.reduce((filtered, context) => {
+      if (context.event.insert_id) {
+        filtered.add(context.event.insert_id);
```

Comment:
In the event schema, why is insert_id nullable? Can we make it non-nullable so that we know that insert id is never going to be null or empty?

Reply:
I think insert_id in the backend is nullable. Only the event_type and user id or device id are required.

Reply:
Yes, it is nullable in the backend but with the current model, it feels like it can be nullable in the SDK too? If it can be null in the SDK, then using this field for filtering might have issues?

Reply:
Will assign the UUID as insert_id if it's null before sending out.